fuel-core/src/database.rs

```diff
@@ -248,3 +260,113 @@ impl InterpreterStorage for Database {
         Ok(block.headers.producer)
     }
 }
+
+#[async_trait]
+impl RelayerDb for Database {
+    /// get validator set for current eth height
+    async fn current_validator_set(&self) -> HashMap<Address, u64> {
+        struct WrapAddress(pub Address);
+        impl From<Vec<u8>> for WrapAddress {
+            fn from(i: Vec<u8>) -> Self {
+                Self(Address::try_from(i.as_ref()).unwrap())
+            }
+        }
+        let mut out = HashMap::new();
+        for diff in self.iter_all::<WrapAddress, u64>(columns::VALIDATOR_SET, None, None, None) {
+            match diff {
+                Ok((address, stake)) => {
+                    out.insert(address.0, stake);
+                }
+                Err(_) => return out,
+            }
+        }
+        out
+    }
+
+    /// get stakes difference between fuel blocks. Return vector of changed (some blocks are not going to have any change)
+    async fn get_validator_set_diff(
+        &self,
+        from_fuel_block: u64,
+        to_fuel_block: Option<u64>,
+    ) -> Vec<(u64, HashMap<Address, u64>)> {
+        let to_fuel_block = if let Some(to_fuel_block) = to_fuel_block {
+            if from_fuel_block > to_fuel_block {
+                return Vec::new();
+            }
+            to_fuel_block
+        } else {
+            u64::MAX
+        };
+        struct Wrapu64BE(pub u64);
+        impl From<Vec<u8>> for Wrapu64BE {
+            fn from(i: Vec<u8>) -> Self {
+                use byteorder::{BigEndian, ReadBytesExt};
+                use std::io::Cursor;
+                let mut i = Cursor::new(i);
+                Self(i.read_u64::<BigEndian>().unwrap_or_default())
+            }
+        }
+        let mut out = Vec::new();
+        for diff in self.iter_all::<Wrapu64BE, HashMap<Address, u64>>(
+            columns::VALIDATOR_SET_DIFFS,
+            None,
+            Some(from_fuel_block.to_be_bytes().to_vec()),
+            None,
+        ) {
+            match diff {
+                Ok((key, diff)) => {
+                    let block = key.0;
+                    if block >= to_fuel_block {
+                        return out;
+                    }
+                    out.push((block, diff))
+                }
+                Err(_) => return out,
+            }
+        }
+        out
+    }
+
+    /// current best block number
+    async fn get_block_height(&self) -> u64 {
```

Comment:
All these methods should return `Result<.., Error>`

Reply:
I handled potential error with panic so it is easier to debug, this kind of db error are not something we can recover from.

Reply:
Is it possible that crashing the node leaves the database in a corrupt state?